common/src/main/java/io/pravega/common/io/serialization/RevisionDataOutputStream.java

```diff
@@ -0,0 +1,372 @@
+            super(outputStream);
+
+            // Pre-allocate 4 bytes so we can write the length later, but remember this position.
+            this.initialPosition = ((RandomOutput) outputStream).size();
+            BitConverter.writeInt(outputStream, 0);
+        }
+
+        @Override
+        public void close() throws IOException {
+            // Calculate the number of bytes written, making sure to exclude the bytes for the length encoding.
+            RandomOutput ros = (RandomOutput) this.out;
+            int length = ros.size() - this.initialPosition - Integer.BYTES;
+
+            // Write the length at the appropriate position.
+            BitConverter.writeInt(ros.subStream(this.initialPosition, Integer.BYTES), length);
+        }
+
+        @Override
+        public OutputStream getBaseStream() {
+            // We need to return an OutputStream that implements RandomOutput, which is our underlying OutputStream (and not us).
+            return this.out;
+        }
+
+        @Override
+        public boolean requiresExplicitLength() {
+            return false;
+        }
+
+        @Override
+        public void length(int length) throws IOException {
+            // Nothing to do.
+        }
+    }
+
+    /**
+     * RevisionDataOutput implementation that writes to a general OutputStream. This will force the caller to explicitly
+     * calculate and declare the length prior to serialization as it cannot be back-filled upon closing.
+     */
+    private static class NonSeekableRevisionDataOutput extends RevisionDataOutputStream {
+        private final OutputStream realStream;
+        private int length;
+
+        NonSeekableRevisionDataOutput(OutputStream outputStream) {
+            super(new LengthRequiredOutputStream());
+            this.realStream = outputStream;
+            this.length = 0;
+        }
+
+        @Override
+        public void close() throws IOException {
+            // We do not want to close the underlying Stream as it may be reused.
+            if (this.length != size()) {
+                // Check if we wrote the number of bytes we declared, otherwise we will have problems upon deserializing.
+                throw new SerializationException(String.format("Unexpected number of bytes written. Declared: %d, written: %d.", this.length, size()));
+            } else if (this.out instanceof LengthRequiredOutputStream) {
+                // We haven't written anything nor declared a length. Write the length prior to exiting.
+                length(0);
+            }
+        }
+
+        @Override
+        public OutputStream getBaseStream() {
+            return this;
+        }
+
+        @Override
+        public boolean requiresExplicitLength() {
+            return true;
+        }
+
+        @Override
+        public void length(int length) throws IOException {
+            if (this.out instanceof LengthRequiredOutputStream) {
```

Comment:
Should this throw if this condition is false?
If not, perhaps we could have a method that returns whether or not the impl wants the length written. That would avoid instanceOf and allow for more that 2 classes.

Reply:
> Should this throw if this condition is false?

I want this to be idempotent. If you call it twice, only the first one takes effect. A second call is ignored.

> perhaps we could have a method that returns whether or not the impl wants the length written

Yes there is: `boolean requiresExplicitLength()`

Reply:
Ok. Then let's make this `if (out.requiresExplicitLength()) {`

Reply:
Oh, no, you misunderstood!

The `RevisionDataOutputStream` has a `requiresExplicitLength` which tells the caller whether length needs to be declared or not. `this.out` is of type `OutputStream` and can be whatever implementation is out there; since I don't have control over that, I can't put the property there.

I did make a slight adjustment though:
- `RevisionDataOutputStream.requiresExplicitLength()` 
    - Will return false if it wraps a `RandomAccessOutputStream` (this was there before; no change here)
    - Will return true if it wraps some other kind of `OutputStream` until the length is set (via `length()`); after that it will return false. This makes reading this code a bit easier.